templates/gsl.install.sh

```diff
@@ -66,187 +163,216 @@ create_directory()
 
 .endmacro # define_create_directory
 .
-.macro define_make_current_directory()
-# make_current_directory jobs [configure_options]
-make_current_directory()
+.macro define_make_functions()
+configure_and_make_and_install()
 {
-    local JOBS=$1
-    shift 1
+    local OPTIONS=("$@")
+    local CONFIGURATION=("${OPTIONS[@]}" "${CONFIGURE_OPTIONS[@]}")    
+
+    configure_options "${CONFIGURATION[@]}"
+    make_jobs --silent
+    make install   
+    configure_links
+}
+
+autogen_and_configure_and_make_and_install()
+{
+    local OPTIONS=("$@")
+    local CONFIGURATION=("${OPTIONS[@]}" "${CONFIGURE_OPTIONS[@]}")   
 
     ./autogen.sh
-    configure_options "$@"
-    make_jobs $JOBS
+    configure_options "${CONFIGURATION[@]}"
+    make_jobs
     make install
     configure_links
 }
 
-.endmacro # define_make_current_directory
-.
-.macro define_make_jobs()
-# make_jobs jobs [make_options]
 make_jobs()
 {
-    local JOBS=$1
-    shift 1
+    local JOBS="$PARALLEL"
 
     # Avoid setting -j1 (causes problems on Travis).
+    local SEQUENTIAL=1
     if [[ $JOBS > $SEQUENTIAL ]]; then
-        make -j$JOBS "$@"
+        make -j"$JOBS" "$@"
     else
         make "$@"
     fi
 }
 
-.endmacro # define_make_jobs
+.endmacro #define_make_functions
 .
 .macro define_utility_functions()
+.   define_enable_exit_on_first_error()
+.   define_disable_exit_on_first_error()
 .   define_configure_links()
+.   define_change_dir_for_build_and_download_and_extract()
+.   define_make_functions()
 .   define_configure_options()
+.   define_make_tests()
 .   define_create_directory()
-.   define_display_functions()
-.   define_initialize_git()
-.   define_make_current_directory()
-.   define_make_jobs()
-.   define_make_tests("false")
 .   define_push_pop_directory()
 .endmacro # define_utility_functions
 .
-.macro define_build_from_tarball()
-# Standard build from tarball.
-build_from_tarball()
+.macro define_builds_from_tarball(install)
+.   define my.install = define_builds_from_tarball.install
+.   if defined(my.install->build(name="icu"))
+.        define_build_from_tarball_icu(my.install)
+.   endif
+.    
+.   if defined(my.install->build(name="zlib"))
+.        define_build_from_tarball_zlib(my.install)
+.   endif
+.
+.   if defined(my.install->build(name="png"))
+.       define_build_from_tarball_png(my.install)
+.   endif
+.   if defined(my.install->build(name="qrencode"))
+.       define_build_from_tarball_qrencode(my.install)
+.   endif
+.
+.   if defined(my.install->build(name="boost"))
+.       define_build_from_tarball_boost(my.install)
+.   endif
+.
+.   if defined(my.install->build(name="zmq"))
+.       define_build_from_tarball_zmq(my.install)
+.   endif
+.
+.   if defined(my.install->build(name="mbedtls"))
+.       define_build_from_tarball_mbedlts(my.install)
+.   endif
+.endmacro #define_builds_from_taball
+.
+.macro define_build_from_tarball_icu(install)
+.   define my.install = define_build_from_tarball_icu.install
+build_from_tarball_icu()
```

Comment:
I think it is debatable whether this is an improvement.  Parameterization allows for minimal manual intervention whereby all parameters across dependencies are colocated as opposed to the fan out of functions whereby previously functions existed only to couple behavior instead of data.

Personally, I prefer the previous approach, especially if it were extended with definitional guards akin to `PARALLEL` whereby alteration of urls or options could be provided via command line.